Use file moves instead of copy & cleanup tmp dir

myles-keough opened the pull request:
This will defer the service restart until the update has already succeeded. If the update fails, delete the temp directory after moving everything back.

Current behavior:
![Current](https://user-images.githubusercontent.com/85883564/158630992-7253260c-4343-40ea-acf1-c5a6409cb171.png)

New behavior:
![Proposed](https://user-images.githubusercontent.com/85883564/158631048-f176b4d3-2c26-437e-8d9a-68bdf5c78b7b.png)

hunting-for-robert-r-meyer commented on updater/update.go:
```diff
@@ -95,17 +95,20 @@ func GetUpdateDetails(extractedFiles []string) (udt ConfigUDT, updates []string,
 func BackupFiles(updates []string, srcDir string) (backupDir string, err error) {
-	backupDir, err = ioutil.TempDir("", "prefix")
+	backupDir, err = CreateTempDir()
+	if err != nil {
```
Shouldn't this fatal exit the function with an error?

myles-keough replied:
I believe the log.Fatal call will exit the program with an error status code. This should be the same result as bubbling the error up.

hunting-for-robert-r-meyer replied:
You are correct. Thanks!

jamesodoherty-huntress commented on updater/update.go:
```diff
@@ -128,7 +131,7 @@ func RollbackFiles(backupDir string, dstDir string) (err error) {
 	for _, f := range files {
 		orig := path.Join(backupDir, path.Base(f.Name()))
 		dstFile := path.Join(dstDir, path.Base(f.Name()))
-		_, err = CopyFile(orig, dstFile)
+		err = MoveFile(orig, dstFile)
 		if nil != err {
 			return err
```
I noticed the code for `BackupFiles` now moves the existing files to the backup directory, causing the originals to be removed from the installation directory. However `RollbackFiles` will only partially roll back files if it encounters any errors moving a file back to the original location.

It would be more robust to use something like multierror (https://github.com/hashicorp/go-multierror) to collect any errors and change this loop to unconditionally attempt MoveFile for all the files being rolled back before returning.

myles-keough replied:
That's a good point. I'll take a shot at ensuring that all files are tried when rolling back.

jamesodoherty-huntress commented on updater/update.go:
```diff
@@ -139,6 +142,14 @@ func RollbackFiles(backupDir string, dstDir string) (err error) {
 
 // InstallUpdate start/stops service and moves the new files into the `installDir`
 func InstallUpdate(udt ConfigUDT, srcFiles []string, installDir string) error {
+	// move the files into the "base directory"
+	for _, f := range srcFiles {
+		err := MoveFile(f, installDir)
+		if err != nil {
+			return err
+		}
+	}
+
```
This is the motivation for moving the files to the backup directory rather than copying them, correct?

myles-keough replied:
Yes. When we move the services, they don't have to be stopped until replacement files have been placed successfully.

jamesodoherty-huntress reviewed (approved):
Looks good.

I added some notes on `RollbackFiles`, but I think we'll find in practice we won't run into the specific situation I hypothesize about in my comment.

tramzee reviewed (commented):
@myles-keough this commit is not supposed to address any of the issues with https://app.shortcut.com/huntress/story/39106/huntress-updater-does-not-cleanly-handle-situations-where-the-temp-directory-is-on-a-different-drive right?

I mention this because I don't think our current implementation of `MoveFiles` will work across file systems. 

I am happy to approve this PR and then I will follow up with a fix to the "cross file system/drive" issue, which I believe to be the problem with the story included above.

myles-keough commented:
@tramzee This does not address the failure to move from a temp directory on another drive. I think the only way to avoid the rename issue is to create the backup directory on the same drive as the agent. Instead of using os.MkdirTemp, perhaps CreateTempDir should create something within the calling exe's dir?

myles-keough commented:
@jamesodoherty-huntress I have added multierror. Please let me know what you think.

@tramzee I have changed CreateTempDir to use the calling exe's directory. This should fix the alternative filesystem move issue.

There is an issue with cleaning up the previous hUpdate.exe. The calling exe cannot delete itself and silently fails. I would appreciate suggestions on how to handle this. Maybe hUpdate should clear out temp directories on run?

jamesodoherty-huntress commented:
@myles-keough The error handling looks good.

The cleanup is a tricky thing to handle. It does seem like the newly installed hUpdate or HuntressUpdater has to somehow clean it up. Always cleaning up the temp directory might not be a bad approach.

AspenScythe commented:
>  I have changed CreateTempDir to use the calling exe's directory. 
What directory would this normally be? Would we be safe to clear it out on run?

myles-keough commented:
@AspenScythe CreateTempDir uses the Window's temp directory. If this is on another filesystem, we can't move files to/from it. I think we should clear all directories with our temporary prefix out on each run (WIP).

This is the same issue [in this ticket](https://app.shortcut.com/huntress/story/39106/huntress-updater-does-not-cleanly-handle-situations-where-the-temp-directory-is-on-a-different-drive).

tramzee commented on updater/handler.go:
```diff
@@ -90,7 +100,7 @@ func UpdateHandler(infoer Infoer, args Args) (int, error) {
 		return EXIT_ERROR, err
 	}
-	defer os.RemoveAll(tmpDir)
+	defer DeleteDirectory(tmpDir)
```
Is there really a good reason to redefine `os.RemoveAll`?

myles-keough replied:
I don't think so - it was just the existing convention. I'm open to changing it.
